# Importing Libraries.

In [1]:
import pandas as pd
import numpy as np
import os

# Setting Path.

In [2]:
path = r'C:\Users\Chaot\02-2024 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Chaot\\02-2024 Instacart Basket Analysis'

# Importing Dataframe.

In [4]:
df_order_prod_merge = pd.read_pickle(os.path.join(path, '02 - Data', 'Prepared Data', 'order_products_merged.pkl'))

In [5]:
df_order_prod_merge

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                    0             0   2539329        1             1   
1                    1             1   2398795        1             2   
2                    2             2    473747        1             3   
3                    3             3   2254736        1             4   
4                    4             4    431534        1             5   
...                ...           ...       ...      ...           ...   
32404854       3360493       3360493   1320836   202557            17   
32404855       3360494       3360494     31526   202557            18   
32404856       3376038       3376038    758936   203436             1   
32404857       3376039       3376039   2745165   203436             2   
32404858       3408036       3408036   3093936   205420             1   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                          2                  8                     NaN   
1                          3                  7                    15.0   
2                          3                 12                    21.0   
3                          4                  7                    29.0   
4                          4                 15                    28.0   
...                      ...                ...                     ...   
32404854                   2                 15                     1.0   
32404855                   5                 11                     3.0   
32404856                   2                  7                     NaN   
32404857                   3                  5                    15.0   
32404858                   4                 14                     NaN   

          product_id  add_to_cart_order  reordered _merge  Unnamed: 0_y  \
0                196                  1          0   both           195   
1                196                  1          1   both           195   
2                196                  1          1   both           195   
3                196                  1          1   both           195   
4                196                  1          1   both           195   
...              ...                ...        ...    ...           ...   
32404854       43553                  2          1   both         43557   
32404855       43553                  2          1   both         43557   
32404856       42338                  4          0   both         42342   
32404857       42338                 16          1   both         42342   
32404858       28818                  8          0   both         28820   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32404854            Orange Energy Shots        64              7     3.7   
32404855            Orange Energy Shots        64              7     3.7   
32404856          Zucchini Chips, Pesto        50             19     6.9   
32404857          Zucchini Chips, Pesto        50             19     6.9   
32404858  Hot Oatmeal Multigrain Raisin       130             14    10.3   

         indicator2  
0              both  
1              both  
2              both  
3              both  
4              both  
...             ...  
32404854       both  
32404855       both  
32404856       both  
32404857       both  
32404858       both  

[32404859 rows x 18 columns]

# Creating subset.

In [6]:
df = df_order_prod_merge[:1000000]

In [7]:
df.shape

(1000000, 18)

In [8]:
df

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                  0             0   2539329        1             1   
1                  1             1   2398795        1             2   
2                  2             2    473747        1             3   
3                  3             3   2254736        1             4   
4                  4             4    431534        1             5   
...              ...           ...       ...      ...           ...   
999995       2731921       2731921   2836489   164627            14   
999996       2731962       2731962   1843600   164632             5   
999997       2731966       2731966    733106   164632             9   
999998       2731974       2731974   1650124   164632            17   
999999       2731979       2731979   3378258   164632            22   

        orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                        2                  8                     NaN   
1                        3                  7                    15.0   
2                        3                 12                    21.0   
3                        4                  7                    29.0   
4                        4                 15                    28.0   
...                    ...                ...                     ...   
999995                   0                 15                     6.0   
999996                   1                 19                     9.0   
999997                   2                 22                    10.0   
999998                   6                 17                    13.0   
999999                   6                 10                    15.0   

        product_id  add_to_cart_order  reordered _merge  Unnamed: 0_y  \
0              196                  1          0   both           195   
1              196                  1          1   both           195   
2              196                  1          1   both           195   
3              196                  1          1   both           195   
4              196                  1          1   both           195   
...            ...                ...        ...    ...           ...   
999995       30489                  2          0   both         30491   
999996       30489                  2          0   both         30491   
999997       30489                  3          1   both         30491   
999998       30489                  1          1   both         30491   
999999       30489                  3          1   both         30491   

           product_name  aisle_id  department_id  prices indicator2  
0                  Soda        77              7     9.0       both  
1                  Soda        77              7     9.0       both  
2                  Soda        77              7     9.0       both  
3                  Soda        77              7     9.0       both  
4                  Soda        77              7     9.0       both  
...                 ...       ...            ...     ...        ...  
999995  Original Hummus        67             20     7.5       both  
999996  Original Hummus        67             20     7.5       both  
999997  Original Hummus        67             20     7.5       both  
999998  Original Hummus        67             20     7.5       both  
999999  Original Hummus        67             20     7.5       both  

[1000000 rows x 18 columns]

In [9]:
#Adding If-Then function.
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [10]:
#Applying the If-Then function to column.
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Chaot\AppData\Local\Temp\ipykernel_31916\3402866153.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
#Finding count of Low/Mid/High range items
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [12]:
#Finding Max amount of Prices column.
df['prices'].max()

14.8

#### Alternative If-Then with loc() function.

In [13]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Chaot\AppData\Local\Temp\ipykernel_31916\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [17]:
df['prices'].max()

14.8

In [18]:
df

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                  0             0   2539329        1             1   
1                  1             1   2398795        1             2   
2                  2             2    473747        1             3   
3                  3             3   2254736        1             4   
4                  4             4    431534        1             5   
...              ...           ...       ...      ...           ...   
999995       2731921       2731921   2836489   164627            14   
999996       2731962       2731962   1843600   164632             5   
999997       2731966       2731966    733106   164632             9   
999998       2731974       2731974   1650124   164632            17   
999999       2731979       2731979   3378258   164632            22   

        orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                        2                  8                     NaN   
1                        3                  7                    15.0   
2                        3                 12                    21.0   
3                        4                  7                    29.0   
4                        4                 15                    28.0   
...                    ...                ...                     ...   
999995                   0                 15                     6.0   
999996                   1                 19                     9.0   
999997                   2                 22                    10.0   
999998                   6                 17                    13.0   
999999                   6                 10                    15.0   

        product_id  add_to_cart_order  reordered _merge  Unnamed: 0_y  \
0              196                  1          0   both           195   
1              196                  1          1   both           195   
2              196                  1          1   both           195   
3              196                  1          1   both           195   
4              196                  1          1   both           195   
...            ...                ...        ...    ...           ...   
999995       30489                  2          0   both         30491   
999996       30489                  2          0   both         30491   
999997       30489                  3          1   both         30491   
999998       30489                  1          1   both         30491   
999999       30489                  3          1   both         30491   

           product_name  aisle_id  department_id  prices indicator2  \
0                  Soda        77              7     9.0       both   
1                  Soda        77              7     9.0       both   
2                  Soda        77              7     9.0       both   
3                  Soda        77              7     9.0       both   
4                  Soda        77              7     9.0       both   
...                 ...       ...            ...     ...        ...   
999995  Original Hummus        67             20     7.5       both   
999996  Original Hummus        67             20     7.5       both   
999997  Original Hummus        67             20     7.5       both   
999998  Original Hummus        67             20     7.5       both   
999999  Original Hummus        67             20     7.5       both   

              price_range    price_range_loc  
0       Mid-range product  Mid-range product  
1       Mid-range product  Mid-range product  
2       Mid-range product  Mid-range product  
3       Mid-range product  Mid-range product  
4       Mid-range product  Mid-range product  
...                   ...                ...  
999995  Mid-range product  Mid-range product  
999996  Mid-range product  Mid-range product  
999997  Mid-range product  Mid-range product  
999998  Mid-range product  Mid-range product  
999999  Mid-range product  Mid-

# Applying to whole dataframe.

In [19]:
#Appling High Range Loc function
df_order_prod_merge.loc[df_order_prod_merge['prices'] > 15, 'price_range'] = 'High-Range Product'

In [20]:
#Appling Mid Range Loc function
df_order_prod_merge.loc[(df_order_prod_merge['prices'] <= 15) & (df_order_prod_merge['prices'] > 5), 'price_range'] = 'Mid-range product' 

In [21]:
#Appling Low Range Loc function
df_order_prod_merge.loc[df_order_prod_merge['prices'] <= 5, 'price_range'] = 'Low-Range Product'

In [22]:
#Checking counts of Price Range High/Med/Low
df_order_prod_merge['price_range'].value_counts(dropna = False)

price_range
Mid-range product     21860860
Low-Range Product     10126321
High-Range Product      417678
Name: count, dtype: int64

# 'For Loop' for Days of the Week.

In [23]:
df_order_prod_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [24]:
result = []

for value in df_order_prod_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [25]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [26]:
df_order_prod_merge['busiest_day'] = result

In [27]:
df_order_prod_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 4.7.2 Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [28]:
#Creating new Result list with 0/1 = Busiest Days, 4/3 = Least Busiest Days, All other numbers = Regular Days
result_2 = []

for value in df_order_prod_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Least busiest days")
  else:
    result_2.append("Regularly busy days")

In [29]:
result_2

['Regularly busy days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Least busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Least busiest days',
 'Regularly busy days',
 'Least busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Least busiest days',
 'Least busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest

# 4.7.3 Check the values of this new column for accuracy. Note any observations in markdown format.

In [30]:
df_order_prod_merge['busiest_days'] = result_2

In [31]:
df_order_prod_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy days    12916111
Busiest days           11864412
Least busiest days      7624336
Name: count, dtype: int64

In [32]:
df_order_prod_merge['orders_day_of_week'].value_counts(dropna =False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

From the above, we can see the total amount of Busiest days is 11,864,412. If we add the count of the orders_days_of_week column for "0" and "1" we get 11,864,412. The numbers "4" and "3" should also equal to the least busiest days with every other day equaling the regularly busy days count.

# 4.7.4 When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”


In [33]:
df_order_prod_merge['order_hour_of_day'].value_counts(dropna =False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [34]:
#We will be dividing Most Orders for the top 8 hours, Average orders for the middle 8 and Fewest Orders for the last 8.
result_3 = []
for value in df_order_prod_merge["order_hour_of_day"]:
    if value in [10,11,14,15,13,12,16,9]:
        result_3.append ("Most Orders")
    elif value in [17,8,18,19,20,7,21,22]:
        result_3.append ("Average Orders")
    else:
        result_3.append ("Fewest Orders")

In [35]:
result_3

['Average Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Ord

In [36]:
df_order_prod_merge['busiest_period_of_day'] = result_3

In [37]:
df_order_prod_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             1             1   2398795        1             2   
2             2             2    473747        1             3   
3             3             3   2254736        1             4   
4             4             4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  ...  Unnamed: 0_y product_name  aisle_id department_id  \
0                  1  ...           195         Soda        77             7   
1                  1  ...           195         Soda        77             7   
2                  1  ...           195         Soda        77             7   
3                  1  ...           195         Soda        77             7   
4                  1  ...           195         Soda        77             7   

   prices  indicator2        price_range     busiest_day         busiest_days  \
0     9.0        both  Mid-range product  Regularly busy  Regularly busy days   
1     9.0        both  Mid-range product  Regularly busy   Least busiest days   
2     9.0        both  Mid-range product  Regularly busy   Least busiest days   
3     9.0        both  Mid-range product      Least busy   Least busiest days   
4     9.0        both  Mid-range product      Least busy   Least busiest days   

  busiest_period_of_day  
0        Average Orders  
1        Average Orders  
2           Most Orders  
3        Average Orders  
4           Most Orders  

[5 rows x 22 columns]

# 4.7.5 Print the frequency for this new column.


In [38]:
df_order_prod_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: count, dtype: int64

# 4.7.7 Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.


In [62]:
df_order_prod_merge.to_pickle(os.path.join(path, '02 - Data', 'Prepared Data', '4.7_order_prod_merge.pkl'))